### PR TITLE
Can register a callback function with HierarchyIntegrator to allow specialization for HierarchyIntegrator::regridHierarchy()

### PR DESCRIPTION
**Pull request by abarret**
### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by drwells** [comment]
Should these functions be called after synchronization?

**Comment by abarret** [comment]
> Should these functions be called after synchronization?

This is designed for data that isn't controlled by the `HierarchyIntegrator`. I put it here so that if you modify any data owned by the integrator, they won't have to synchronize it. But at the same time this means if you want to _use_ the data, you need to synchronize it first. I think the best thing is to make it clear when this function is being called.

File: ibtk/include/ibtk/HierarchyIntegrator.h [hunk 1]
```diff
@@ -493,6 +493,24 @@ class HierarchyIntegrator : public SAMRAI::mesh::StandardTagAndInitStrategy<NDIM
...
+                 double data_time,
+                 bool initial_time,
+                 void* ctx);
+
+    /*!
+     * Register a callback function to enable further specialization of regridHierarchy().
+     *
+     * \note This function will be called after all data owned by the hierarchy is initialized and set, but before the
+     * data is synchronized. If data owned by the integrator is used in this function, it should be synchronized prior
+     * to use.
```
Comment (drwells) [inline ibtk/include/ibtk/HierarchyIntegrator.h]: From the other conversation: if this is called prior to synchronization but a user needs synchronized data, then how should they do it?

Comment (abarret) [reply ibtk/include/ibtk/HierarchyIntegrator.h]: Using an appropriate coarsening operation. I've updated the comment.

**Comment by drwells** [comment]
@abarret Do you think its necessary to add a new test? I'm not sure if its worth waiting until we use this or not. Additionally, can you run the test suite locally?

**Comment by abarret** [comment]
> Do you think its necessary to add a new test? I'm not sure if its worth waiting until we use this or not.

I've been using this patch for the past several weeks. Now that CIB is fixed, we need this patch for related work. If you approve of the updated documentation, I propose we merge now, and I'll commit to writing tests for all the `HierarchyIntegrator` callback functions soon.

> Additionally, can you run the test suite locally?

Yes. I'm running it now. Again, I've been using this patch for several weeks now so it _should_ be fine.

**Comment by drwells** [comment]
Sounds good but lets run the test suite just to be sure. Would you please merge this correctly once the test suite is done and there are no unexpected failures?

**Comment by abarret** [comment]
The test suite passes so I will merge.

**Pull request merged**